libavpipe/src/avpipe_copy_mpegts.c

```diff
@@ -0,0 +1,389 @@
+ */
+
+#include <libavutil/log.h>
+#include "libavutil/audio_fifo.h"
+#include <libswscale/swscale.h>
+#include <libavutil/imgutils.h>
+#include <libavutil/display.h>
+
+#include "avpipe_xc.h"
+#include "avpipe_utils.h"
+#include "avpipe_format.h"
+#include "avpipe_io.h"
+#include "avpipe_copy_mpegts.h"
+#include "elv_log.h"
+#include "elv_time.h"
+#include "url_parser.h"
+#include "avpipe_version.h"
+
+#include <stdio.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/uio.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <pthread.h>
+
+
+int
+copy_mpegts_init(
+    xctx_t *xctx)
+{
+    // Nothing to do
+    return eav_success;
+}
+
+static int
+copy_mpegts_set_encoder_options(
+    coderctx_t *encoder_context,
+    coderctx_t *decoder_context,
+    xcparams_t *params,
+    int stream_index,
+    int timebase)
+{
+    if (timebase <= 0) {
+        elv_err("Setting encoder options failed, invalid timebase=%d (check encoding params), url=%s",
+            timebase, params->url);
+        return eav_timebase;
+    }
+
+    int64_t seg_duration_ts = 0;
+    float seg_duration = 0;
+
+    /* Precalculate seg_duration_ts based on seg_duration if seg_duration is set */
+    if (params->seg_duration) {
+        seg_duration = atof(params->seg_duration);
+        if (stream_index == decoder_context->video_stream_index)
+            timebase = calc_timebase(params, 1, timebase);
+        seg_duration_ts = seg_duration * timebase;
+    }
+    if (params->video_seg_duration_ts > 0)
+        seg_duration_ts = params->video_seg_duration_ts;
+
+    /* If audio_seg_duration_ts is not set, set it now */
+    if (params->audio_seg_duration_ts <= 0)
+        params->audio_seg_duration_ts = seg_duration_ts;
+
+    av_opt_set_int(encoder_context->format_context->priv_data, "segment_duration_ts", seg_duration_ts, 0);
+
+    /* If video_seg_duration_ts is not set, set it now */
+    if (params->video_seg_duration_ts <= 0)
+        params->video_seg_duration_ts = seg_duration_ts;
+
+    elv_dbg("setting \"fmp4-segment\" video segment_time to %s, seg_duration_ts=%"PRId64", url=%s",
+    params->seg_duration, seg_duration_ts, params->url);
+
+    return eav_success;
+}
+
+static int
+copy_mpegts_prepare_video_encoder(
+    coderctx_t *encoder_context,
+    coderctx_t *decoder_context,
+    xcparams_t *params)
+{
+    int rc = 0;
+    int index = decoder_context->video_stream_index;
+
+    if (index < 0) {
+        elv_dbg("No video stream detected by decoder.");
+        return eav_stream_index;
+    }
+
+    encoder_context->video_stream_index = index;
+    encoder_context->video_last_dts = AV_NOPTS_VALUE;
+    encoder_context->stream[index] = avformat_new_stream(encoder_context->format_context, NULL);
+    encoder_context->codec[index] = avcodec_find_encoder_by_name(params->ecodec);
+
+    /* Custom output buffer */
+    encoder_context->format_context->io_open = elv_io_open;
```

Comment:
This is one place where we could just take a different path for the mpegts output. It's probably the earliest one.